🔍 Investigate and fix chat template initialization error during training

Assigned to: @claude 

🧩 Issue Summary

Running make train-accelerate fails with a ValueError during model setup, specifically when applying the chat template through trl.models.utils.setup_chat_format. The traceback indicates that the tokenizer already contains a chat template, causing the setup process to crash.

📄 Logs

```
[2025-10-16 13:22:48][src.model_setup][ERROR] - Failed to setup chat format: Chat template is already added to the tokenizer. If you want to overwrite it, please set it to None
Error executing job with overrides: []
Traceback (most recent call last):
  File "/home/chrisjcc/text-to-sql-finetuning/scripts/train.py", line 61, in main
    model, tokenizer, _ = initialize_model_for_training(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chrisjcc/text-to-sql-finetuning/src/model_setup.py", line 383, in initialize_model_for_training
    model, tokenizer = setup.setup_for_chat(model, tokenizer)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chrisjcc/text-to-sql-finetuning/src/model_setup.py", line 118, in setup_for_chat
    model, tokenizer = setup_chat_format(model, tokenizer)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chrisjcc/miniconda3/envs/llm_env/lib/python3.11/site-packages/trl/models/utils.py", line 123, in setup_chat_format
    raise ValueError(
ValueError: Chat template is already added to the tokenizer. If you want to overwrite it, please set it to None

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
Traceback (most recent call last):
  File "/home/chrisjcc/miniconda3/envs/llm_env/bin/accelerate", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/chrisjcc/miniconda3/envs/llm_env/lib/python3.11/site-packages/accelerate/commands/accelerate_cli.py", line 50, in main
    args.func(args)
  File "/home/chrisjcc/miniconda3/envs/llm_env/lib/python3.11/site-packages/accelerate/commands/launch.py", line 1235, in launch_command
    simple_launcher(args)
  File "/home/chrisjcc/miniconda3/envs/llm_env/lib/python3.11/site-packages/accelerate/commands/launch.py", line 823, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['/home/chrisjcc/miniconda3/envs/llm_env/bin/python3', '-m', 'scripts.train']' returned non-zero exit status 1.
⚠ Training with accelerate failed.
make: *** [Makefile:50: train-accelerate] Error 1
```

⚙️ Additional Context

- Command executed: make train-accelerate
- Model: meta-llama/Llama-3.1-8B-Instruct
- Hardware: NVIDIA H200 (150 GB)
- Warnings before failure:
    - setup_chat_format is deprecated; should use clone_chat_template (per TRL v0.26.0 notice)
    - The tokenizer appears to already include a chat template, possibly from the pretrained model.
    - Deprecation: torch_dtype should be replaced by dtype.

🧠 Likely Root Cause

setup_chat_format() attempts to re-apply a chat template even though the Llama-3.1-8B-Instruct tokenizer already includes one. TRL’s newer models now embed chat templates natively, making manual setup redundant or incompatible.

✅ Tasks

1. Detect and skip redundant chat setup:
In src/model_setup.py, modify setup_for_chat() to check if the tokenizer already includes a chat template before calling setup_chat_format().
```python
if hasattr(tokenizer, "chat_template") and tokenizer.chat_template is not None:
    logger.warning("Chat template already present, skipping setup_chat_format()")
else:
    model, tokenizer = setup_chat_format(model, tokenizer)

```
2. Transition to clone_chat_template():
Replace deprecated setup_chat_format() with clone_chat_template() following the TRL ≥0.26.0 API.

3. Test and confirm successful initialization:
- Run make train-accelerate again.
- Verify that the training proceeds beyond the model setup phase.
- Ensure chat formatting remains consistent in sample generations.
- Add a config flag in Hydra to explicitly control chat setup, e.g. (figure out where to include it e.g. evaluation.yaml and or training.yaml):
```
model:
  setup_chat_format: false
```
To support Instruct models and base models.

🧾 Notes

The earlier warnings about accelerate defaults and pydantic attributes are non-fatal but should be cleaned up later for a smoother log output. The git credential and HF_TOKEN warnings are unrelated to the failure and do not block training.

--- a/config/hf/hf.yaml
+++ b/config/hf/hf.yaml
@@ -5,6 +5,12 @@ model_id: meta-llama/Llama-3.1-8B-Instruct
 token: ${oc.env:HF_TOKEN,null}
 username: chrisjcc
 
+# Model setup configuration
+# Set setup_chat_format to false for Instruct models that already have chat templates
+# Set to true for base models that need chat template setup
+setup_chat_format: false  # Llama-3.1-8B-Instruct already has a chat template
+force_chat_setup: false    # Set to true to force re-applying chat template (not recommended)
+
 # Upload configuration
 upload:
   # What to upload

--- a/scripts/train.py
+++ b/scripts/train.py
@@ -58,10 +58,15 @@ def main(cfg: DictConfig):
 
     # Initialize model and tokenizer
     use_flash_attention = cfg.training.get("use_flash_attention", True)
+    setup_chat_format = cfg.hf.get("setup_chat_format", True)
+    force_chat_setup = cfg.hf.get("force_chat_setup", False)
+
     model, tokenizer, _ = initialize_model_for_training(
         model_id=cfg.hf.model_id,
         use_flash_attention=use_flash_attention,
         max_seq_length=cfg.training.max_seq_length,
+        setup_chat_format=setup_chat_format,
+        force_chat_setup=force_chat_setup,
     )
 
     # Configure LoRA if applicable

--- a/src/model_setup.py
+++ b/src/model_setup.py
@@ -101,7 +101,8 @@ class ModelSetup:
     @staticmethod
     def setup_for_chat(
         model: PreTrainedModel,
-        tokenizer: PreTrainedTokenizer
+        tokenizer: PreTrainedTokenizer,
+        force_setup: bool = False
     ) -> Tuple[PreTrainedModel, PreTrainedTokenizer]:
         """
         Setup model and tokenizer for chat format.
@@ -109,12 +110,27 @@ class ModelSetup:
         Args:
             model: Pre-trained model
             tokenizer: Pre-trained tokenizer
+            force_setup: If True, force setup even if chat template exists (will set to None first)
 
         Returns:
             Tuple of (configured model, configured tokenizer)
         """
         logger.info("Setting up chat format for model and tokenizer")
+
+        # Check if tokenizer already has a chat template
+        has_chat_template = hasattr(tokenizer, "chat_template") and tokenizer.chat_template is not None
+
+        if has_chat_template:
+            if force_setup:
+                logger.warning("Chat template already present, but force_setup=True. Setting to None and re-applying.")
+                tokenizer.chat_template = None
+            else:
+                logger.info("Chat template already present in tokenizer, skipping setup_chat_format()")
+                logger.info("This is expected for Instruct models like Llama-3.1-8B-Instruct")
+                return model, tokenizer
+
         try:
+            logger.info("Applying chat format setup...")
             model, tokenizer = setup_chat_format(model, tokenizer)
             logger.info("Chat format setup completed")
             return model, tokenizer
@@ -356,7 +372,9 @@ def initialize_model_for_training(
     model_id: str,
     use_flash_attention: bool = True,
     max_seq_length: int = 2048,
-    lora_config: LoraConfig = None
+    lora_config: LoraConfig = None,
+    setup_chat_format: bool = True,
+    force_chat_setup: bool = False
 ) -> Tuple[PreTrainedModel, PreTrainedTokenizer, LoraConfig]:
     """
     Convenience function to initialize model for training.
@@ -366,6 +384,9 @@ def initialize_model_for_training(
         use_flash_attention: Whether to attempt Flash Attention 2
         max_seq_length: Maximum sequence length for the model
         lora_config: Optional pre-configured LoRA config
+        setup_chat_format: Whether to setup chat format (default: True)
+                          Set to False for models that already have chat templates (e.g., Instruct models)
+        force_chat_setup: If True, force chat setup even if template exists (default: False)
 
     Returns:
         Tuple of (model, tokenizer, lora_config)
@@ -379,8 +400,11 @@ def initialize_model_for_training(
         max_seq_length
     )
 
-    # Setup chat format
-    model, tokenizer = setup.setup_for_chat(model, tokenizer)
+    # Setup chat format (conditionally)
+    if setup_chat_format:
+        model, tokenizer = setup.setup_for_chat(model, tokenizer, force_setup=force_chat_setup)
+    else:
+        logger.info("Skipping chat format setup (setup_chat_format=False)")
 
     # Create LoRA config if not provided
     if lora_config is None:

